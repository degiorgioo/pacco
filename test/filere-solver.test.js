const { expect, beforeEach } = require("@jest/globals");
const path = require("path");
const fs = require("fs");
const {
  readPackageJson,
  fileResolverErrorMessages,
} = require("../src/file-resolver");

/**
 * Testing contexts.
 */
const testingBasePath = path.join(__dirname, "fixtures", "packageJson");
const correctContext = path.join(testingBasePath, "correct");
const missingContext = path.join(testingBasePath, "missing");
const erronousContext = path.join(testingBasePath, "error");

/**
 * Mocks.
 */
const processCwdMock = jest.spyOn(process, "cwd");

beforeEach(() => {
  jest.resetAllMocks();
});

afterEach(() => {
  jest.resetAllMocks();
});

describe("fileresolver.js tests", () => {
  it("should read a correct package.json file correctly without errors", () => {
    // given
    processCwdMock.mockReturnValue(correctContext);

    // when
    const result = readPackageJson();

    // then
    expect(result).toMatchSnapshot();
  });

  it("should throw when package.json file is missing", () => {
    // givens
    processCwdMock.mockReturnValue(missingContext);

    // when
    // then
    expect(readPackageJson).toThrowError(
      fileResolverErrorMessages.noPackageJsonFileFound
    );
  });

  it("should throw when package.json could not be read", () => {
    // given
    const fileSystemReadFileSyncMock = jest.spyOn(fs, "readFileSync");
    processCwdMock.mockReturnValue(erronousContext);

    fileSystemReadFileSyncMock.mockImplementation(() => {
      throw new Error();
    });

    // when
    // then
    expect(readPackageJson).toThrowError(
      fileResolverErrorMessages.packageJsonFileReadError
    );
  });

  it("should throw when package.json could not be parsed", () => {
    // given
    const jsonParseMock = jest.spyOn(JSON, "parse");
    processCwdMock.mockReturnValue(erronousContext);

    jsonParseMock.mockImplementation(() => {
      throw new Error();
    });

    // when
    // then
    expect(readPackageJson).toThrowError(
      fileResolverErrorMessages.packageJsonParsingError
    );
  });

  it("should throw when package.json does not have scripts defined", () => {
    // given
    const jsonParseMock = jest.spyOn(JSON, "parse");
    processCwdMock.mockReturnValue(erronousContext);
    jsonParseMock.mockImplementation(() => ({}));

    // when
    // then
    expect(readPackageJson).toThrowError(
      fileResolverErrorMessages.packageJsonNoScriptsError
    );
  });
});
