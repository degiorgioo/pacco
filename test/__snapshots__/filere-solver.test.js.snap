// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fileresolver.js tests should read a correct package.json file correctly without errors 1`] = `
Object {
  "dependencies": Object {
    "cli-color": "^2.0.0",
    "inquirer": "^8.0.0",
    "inquirer-search-list": "^1.2.6",
  },
  "description": "Helper to run package.json scripts.",
  "devDependencies": Object {
    "jest": "^26.6.3",
  },
  "main": "index.js",
  "name": "@vai/correct",
  "scripts": Object {
    "build": "echo build",
    "start": "echo start",
    "test-watch": "jest --watchAll",
  },
  "version": "1.0.0",
}
`;
